### PR TITLE
Schedule owned-object transactions for execution from consensus validator to reduce latency

### PR DESCRIPTION
**Pull request by mystenmark**
This exposed some weird test issues (reconfig tests could end up taking forever due to ever-increasing QD retry delays), hence the test and QD changes.

### CODE REVIEW

File: crates/sui-core/src/consensus_validator.rs [hunk 1]
```diff
@@ -127,16 +137,27 @@ impl TransactionValidator for SuiTxValidator {
         self.metrics
             .checkpoint_signatures_verified
             .inc_by(ckpt_count as u64);
-        Ok(())
-
         // todo - we should un-comment line below once we have a way to revert those transactions at the end of epoch
```
Comment (andll) [inline crates/sui-core/src/consensus_validator.rs:140]: Should we remove this comment as well?

**Review by andll — APPROVED** [review]
This looks pretty clean.

Are there any meaningful perf numbers we could measure for this change? (e.g. client latency improvement)

File: crates/sui-core/src/consensus_validator.rs [hunk 2]
```diff
@@ -127,16 +137,27 @@ impl TransactionValidator for SuiTxValidator {
...
         // via Narwhal output.
-        // self.transaction_manager
-        //     .enqueue_certificates(owned_tx_certs, &self.epoch_store)
-        //     .wrap_err("Failed to schedule certificates for execution")
+        let reconfiguration_lock = epoch_store.get_reconfig_state_read_lock_guard();
+        if reconfiguration_lock.should_accept_user_certs() {
+            for cert in owned_tx_certs.iter() {
+                let transaction =
+                    ConsensusTransaction::new_certificate_message(&self.name, cert.clone().into());
+                epoch_store.insert_pending_consensus_transactions(
```
Comment (andll) [inline crates/sui-core/src/consensus_validator.rs:150]: Maybe we could change this API to insert all transactions in one batch?

File: crates/sui-core/src/consensus_validator.rs [hunk 3]
```diff
@@ -127,16 +137,27 @@ impl TransactionValidator for SuiTxValidator {
...
-
         // todo - we should un-comment line below once we have a way to revert those transactions at the end of epoch
         // all certificates had valid signatures, schedule them for execution prior to sequencing
         // which is unnecessary for owned object transactions.
         // It is unnecessary to write to pending_certificates table because the certs will be written
         // via Narwhal output.
-        // self.transaction_manager
-        //     .enqueue_certificates(owned_tx_certs, &self.epoch_store)
-        //     .wrap_err("Failed to schedule certificates for execution")
+        let reconfiguration_lock = epoch_store.get_reconfig_state_read_lock_guard();
```
Comment (longbowlu) [inline crates/sui-core/src/consensus_validator.rs:145]: is the comment above still relevant?

File: crates/sui-core/src/consensus_validator.rs (hunk 2 above)
Comment (mystenmark) [reply crates/sui-core/src/consensus_validator.rs:150]: good idea, use a single write batch...

**Comment by mystenmark** [comment]
> This looks pretty clean.
> 
> Are there any meaningful perf numbers we could measure for this change? (e.g. client latency improvement)

I don't think we'll likely be able to measure any difference except on organic traffic, because the only case this helps with is when shared objects are waiting on unsequenced dependencies. I'm hoping we'll see a slight drop in p95 shared txn latency.

File: crates/sui-core/src/consensus_validator.rs (hunk 1 above)
Comment (mystenmark) [reply crates/sui-core/src/consensus_validator.rs]: done

File: crates/sui-core/src/consensus_validator.rs (hunk 2 above)
Comment (mystenmark) [reply crates/sui-core/src/consensus_validator.rs]: done

File: crates/sui-core/src/consensus_validator.rs (hunk 3 above)
Comment (mystenmark) [reply crates/sui-core/src/consensus_validator.rs:141]: removed

File: crates/sui-core/src/quorum_driver/mod.rs [hunk 4]
```diff
@@ -143,8 +146,12 @@ impl<A: Clone> QuorumDriver<A> {
             );
             return Ok(());
         }
-        let next_retry_after =
-            Instant::now() + Duration::from_millis(200 * u64::pow(2, old_retry_times.into()));
+        let mut next_retry_after = Duration::from_millis(200 * u64::pow(2, old_retry_times.into()));
```
Comment (mwtian) [inline crates/sui-core/src/quorum_driver/mod.rs:149]: Always having a `max_retry_delay` of 10s ~ 60s seems good.

**Review by mwtian — APPROVED** [review]

File: crates/sui-core/src/consensus_validator.rs [hunk 5]
```diff
@@ -78,19 +83,24 @@ impl TransactionValidator for SuiTxValidator {
...
-                    //     // have had their signatures verified.
-                    //     owned_tx_certs.push(VerifiedCertificate::new_unchecked(*certificate));
-                    // }
+                    cert_batch.push(*certificate.clone());
+
+                    if !certificate.contains_shared_object()
+                        && !epoch_store.is_tx_cert_consensus_message_processed(&certificate)?
+                    {
+                        // new_unchecked safety: we do not use the certs in this list until all
+                        // have had their signatures verified.
```
Comment (longbowlu) [inline crates/sui-core/src/consensus_validator.rs:101]: can we update the comment here to call out the sigs are verified for `cert_batch` below?

File: crates/sui-core/src/authority/authority_per_epoch_store.rs [hunk 6]
```diff
@@ -964,22 +964,28 @@ impl AuthorityPerEpochStore {
...
+        let mut batch = self.tables.pending_consensus_transactions.batch();
+        batch.insert_batch(
+            &self.tables.pending_consensus_transactions,
+            transactions.iter().map(|t| (t.key(), t)),
+        )?;
+        batch.write()?;
+
+        for transaction in transactions {
+            if let ConsensusTransactionKind::UserTransaction(cert) = &transaction.kind {
+                let state = lock.expect("Must pass reconfiguration lock when storing certificate");
```
Comment (longbowlu) [inline crates/sui-core/src/authority/authority_per_epoch_store.rs:979]: why is the `lock` in the function signature option?

**Review by longbowlu — APPROVED** [review]

**Pull request merged**